File: source/cpp/100. Same Tree.cpp
#include <cassert>
#include <iostream>
#include <string>


// uncomment to disable assert()
// #define NDEBUG

// Use (void) to silence unused warnings.
#define assertm(exp, msg) assert(((void)msg, exp))



// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr && q == nullptr)
            return true;
        
        if (p == nullptr || q == nullptr || p->val != q->val)
            return false;

        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};


int main()
{
    Solution solution;

    TreeNode* p = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    TreeNode* q = new TreeNode(1, new TreeNode(2), new TreeNode(3));

    bool result = solution.isSameTree(p, q);

    assertm(result == true, "Result is not true.");

    std::cout << "All tests passed." << std::endl;
}

File: source/python/100. Same Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int = 0, left: 'TreeNode' | None = None, right: 'TreeNode' | None = None):
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


class Solution:
    def isSameTree(self, p: TreeNode | None, q: TreeNode | None) -> bool:
        if p is None and q is None:
            return True
        if p is None or q is None or p.val != q.val:
            return False

        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)


if __name__ == '__main__':
    solution = Solution()

    assert solution.isSameTree(None, None) is True
    assert solution.isSameTree(None, TreeNode(1)) is False
    assert solution.isSameTree(TreeNode(1), None) is False

    assert solution.isSameTree(TreeNode(1, TreeNode(2), TreeNode(
        3)), TreeNode(1, TreeNode(2), TreeNode(3))) is True
    assert solution.isSameTree(TreeNode(1, TreeNode(
        2), None), TreeNode(1, None, TreeNode(2))) is False
    assert solution.isSameTree(TreeNode(1, TreeNode(2), TreeNode(
        1)), TreeNode(1, TreeNode(1), TreeNode(2))) is False

    print("All tests passed.")
